lib/matplotlib/image.py

```diff
@@ -589,11 +589,11 @@ def set_data(self, A):
             self._A = cbook.safe_masked_invalid(A, copy=True)
 
         if (self._A.dtype != np.uint8 and
-                not np.can_cast(self._A.dtype, float)):
-            raise TypeError("Image data can not convert to float")
+                not np.can_cast(self._A.dtype, float, "same_kind")):
+            raise TypeError("Image data cannot be converted to float")
 
-        if (self._A.ndim not in (2, 3) or
-                (self._A.ndim == 3 and self._A.shape[-1] not in (3, 4))):
+        if not (self._A.ndim == 2
+                or self._A.ndim == 3 and self._A.shape[-1] in [3, 4]):
```

Comment:
the boolean structure looks correct to me
i think the newline already helps with priority